File: payment-service/src/events/event-envelope.ts
import { randomUUID } from 'crypto';

/**
 * Enveloppe partagée de tous les événements FoodFlow (convention polyglotte,
 * répliquée à l'identique côté order-service en Java).
 *
 * - `eventId`     : identifiant unique DU message (base de l'idempotence)
 * - `eventType`   : type de l'événement, ex. "PaymentAuthorized"
 * - `occurredAt`  : date ISO-8601 où le fait s'est produit
 * - `aggregateId` : identifiant de l'agrégat (l'orderId) — aussi la clé Kafka
 * - `payload`     : données métier propres au type d'événement
 */
export interface EventEnvelope<T> {
  eventId: string;
  eventType: string;
  occurredAt: string;
  aggregateId: string;
  payload: T;
}

/**
 * Fabrique un événement en générant automatiquement son `eventId`
 * et son `occurredAt`.
 */
export function createEvent<T>(
  eventType: string,
  aggregateId: string,
  payload: T,
): EventEnvelope<T> {
  return {
    eventId: randomUUID(),
    eventType,
    occurredAt: new Date().toISOString(),
    aggregateId,
    payload,
  };
}

File: order-service/src/main/java/fr/esgi/foodflow/order_service/event/EventEnvelope.java
package fr.esgi.foodflow.order_service.event;

import java.time.Instant;
import java.util.UUID;

/**
 * Enveloppe partagée de tous les événements FoodFlow (convention polyglotte,
 * répliquée à l'identique côté payment-service en TypeScript).
 *
 * @param eventId     identifiant unique DU message (base de l'idempotence)
 * @param eventType   type de l'événement, ex. "OrderCreated"
 * @param occurredAt  instant où le fait s'est produit
 * @param aggregateId identifiant de l'agrégat (l'orderId) — aussi la clé Kafka
 * @param payload     données métier propres au type d'événement
 */
public record EventEnvelope<T>(
        UUID eventId,
        String eventType,
        Instant occurredAt,
        UUID aggregateId,
        T payload) {

    /**
     * Fabrique un événement en générant automatiquement son {@code eventId}
     * et son {@code occurredAt}.
     */
    public static <T> EventEnvelope<T> of(String eventType, UUID aggregateId, T payload) {
        return new EventEnvelope<>(UUID.randomUUID(), eventType, Instant.now(), aggregateId, payload);
    }
}
